mobile-widgets/qmlinterface.cpp

```diff
@@ -95,8 +95,15 @@ void QMLInterface::setup(QQmlContext *ct)
 	diveSummary::summaryCalculation(0, 3);
 }
 
+static QStringList diveSummaryTextCached;
+
 void QMLInterface::summaryCalculation(int primaryPeriod, int secondaryPeriod)
 {
-	diveSummary::summaryCalculation(primaryPeriod, secondaryPeriod);
-	emit diveSummaryTextChanged(diveSummary::diveSummaryText);
+	diveSummaryTextCached = diveSummary::summaryCalculation(primaryPeriod, secondaryPeriod);
```

Comment:
it seems we are copying the stringiest here, is that really needed ?

Reply:
> it seems we are copying the stringlist here

Perhaps surprisingly we aren't. This is for two reasons:
1) Qt's containers use CopyOnWrite semantics. An assignment of two containers does no deep copy, but simply increases a reference count. A deep copy will only be performed if you modify one of the containers. This has some nasty implications, by the way. You will even find recommendations to pass QString, etc. by value in signals to avoid subtle time-of-life or concurrent-access bugs. Matter of fact, we did have horrible-to-debug issues because the QModelIndex we got in a signal suddenly changed under our feet. Lesson learned: Qt's proxy models aren't reentrant.
2) Even if this were standard non-COW containers, which I strongly prefer, this would be subject to return value optimization and copy elision: https://en.wikipedia.org/wiki/Copy_elision#Return_value_optimization
Take for example the code
```
#include <cstdio>

struct X {
	X() { printf("Construct\n"); }
	X(const X &) { printf("Copy construct\n"); }
	X &operator=(const X &) { printf("Copy assign\n"); return *this; }
	X &operator=(X &&) { printf("Move\n"); return *this; }
};

X f()
{
	return X();
}

int main()
{
	X x1 = f();
	printf("--\n");
	x1 = f();
}
```
This outputs:
```
Construct
--
Construct
Move
```
There is not a single copy (by assignment or construction) performed. A move operation of a container is usually more-or-less free. I could imagine that on modern CPUs it is even more effective than the indirection caused by references! But don't quote me on that.

Reply:
thanks for the explanation, then the only remaining problem is that QMLInterface is no longer a pure interface, but carries data.

Reply:
I remember our conversation that you felt that it being a pure interface was important, but if asked I don't think I could reconstruct that argument... in other words, would you please re-explain that to me (I know you already did... I am simply very slow when it comes to internalizing object oriented programming subtleties).

Reply:
> [...] pure interface  [...] object oriented programming  subtleties.

Sorry for being that annoying person, but `QMLInterface` is *not* an interface class in the sense that you would typically use the word in OO programming (especially Java and other languages that don't have multiple inheritance).

In OO context, an interface *defines* methods that a class has to provide to fulfill that interface but it does not *implement* them. Classes can implement one or more interfaces by providing the appropriate functions.

In C++ terms, an interface class would consist only of pure virtual functions, i.e. function of the type
```
   virtual void f() = 0;
```
`QMLInterface` is, in contrast, very much an *implementation* of a number of functions.
Of course that all has nothing to do with your question, it's just clarification of terminology. Also note that I expressly do *not* suggest to rename the thing.

Reply:
please QMLInterface does not contain implementations, but merely calls to classes/functions that contain the actual implementation, in that respect it is a pure interface. This type of interfaces are often found as API, allowing a library to only expose the methods it wants to expose. But please this is not a terminology discussion, making the right choice is far more  important, because the choice have consequences. 

My concern is to avoid another monolith like qmlmanager, if we start allowing implementation code in qmlinterface, the simplicity will soon be gone, right now it serves one purpose and one purpose only, to be interface to QML, it hold the Q_PROPERTY etc. definitions.

If we allow implementations we can start to move quite a lot of mobile specific code from e.g. diveplannerpointsmodel into qmlinterface and then we end having a mixture of actual implemention code and interface calls, which soon becomes interesting to maintain.

The goal was to simplify the QML interface and make it clear what code is interface. If we move away from that we should move qmlinterface into qmlmanager to at least avoid building 2 momoliths.

Reply:
Please allow me to try to insert myself as a translator here :-)

Ignoring for a moment the correct definition of what an "interface class" is...

I believe what Jan is talking about makes sense. Having a "clean class" (to invent a new term) between the UI and the backend seems useful. This exposes the methods and signals and accessors that the QML code needs, and simply then calls shared code (or sometimes mobile specific code that isn't UI related) to collect the data, but doesn't hold any of that data itself.

Can I claim that this is absolutely necessary for Subsurface-mobile to be created?
No, I don't think I can.
Can I claim that I understand that this is a useful construct to make things cleaner and easier to understand?
Yeah, possibly.

Maybe this helped move the conversation forward :-)

If not, feel free to curse me and move on arguing :-(